File: packages/react-native-vision-camera/ios/Hybrid Objects/HybridCameraController.swift
///
/// HybridCameraController.swift
/// VisionCamera
/// Copyright © 2025 Marc Rousavy @ Margelo
///

import AVFoundation
import Foundation
import NitroModules

final class HybridCameraController: HybridCameraControllerSpec, NativeCameraController {
  let device: any HybridCameraDeviceSpec
  let queue: DispatchQueue
  let captureDevice: AVCaptureDevice
  private var activeMeteringTask: MeteringTask? = nil

  init(device: any HybridCameraDeviceSpec, queue: DispatchQueue) throws {
    guard let hybridDevice = device as? HybridCameraDevice else {
      throw RuntimeError.error(
        withMessage: "Device \"\(device)\" is not of type `HybridCameraDevice`!")
    }
    self.captureDevice = hybridDevice.device
    self.device = device
    // TODO: Use our own queue here?
    self.queue = queue
  }

  var isConnected: Bool {
    return captureDevice.isConnected
  }

  var isSuspended: Bool {
    return captureDevice.isSuspended
  }

  var isUsedByAnotherApp: Bool {
    #if os(macOS)
      return captureDevice.isInUseByAnotherApplication
    #else
      return false
    #endif
  }

  var focusMode: FocusMode {
    return FocusMode(mode: captureDevice.focusMode)
  }

  var exposureMode: ExposureMode {
    return ExposureMode(mode: captureDevice.exposureMode)
  }

  var whiteBalanceMode: WhiteBalanceMode {
    return WhiteBalanceMode(mode: captureDevice.whiteBalanceMode)
  }

  var whiteBalanceGains: WhiteBalanceGains {
    let gains = captureDevice.deviceWhiteBalanceGains
    return WhiteBalanceGains(
      redGain: Double(gains.redGain),
      blueGain: Double(gains.blueGain),
      greenGain: Double(gains.greenGain))
  }

  var isLowLightBoostEnabled: Bool {
    return captureDevice.automaticallyEnablesLowLightBoostWhenAvailable
  }

  var isSmoothAutoFocusEnabled: Bool {
    return captureDevice.isSmoothAutoFocusEnabled
  }

  var isDistortionCorrectionEnabled: Bool {
    return captureDevice.isGeometricDistortionCorrectionEnabled
  }

  var zoom: Double {
    return captureDevice.videoZoomFactor
  }

  var minZoom: Double {
    return captureDevice.minAvailableVideoZoomFactor
  }

  var maxZoom: Double {
    return captureDevice.maxAvailableVideoZoomFactor
  }

  var displayableZoomFactor: Double {
    guard #available(iOS 18.0, *) else {
      return zoom
    }
    return zoom * captureDevice.displayVideoZoomFactorMultiplier
  }

  var torchStrength: Double {
    return Double(captureDevice.torchLevel)
  }

  var torchMode: TorchMode {
    return TorchMode(mode: captureDevice.torchMode)
  }

  var lensPosition: Double {
    return Double(captureDevice.lensPosition)
  }

  var exposureDuration: Double {
    return captureDevice.exposureDuration.seconds
  }

  var activeMaxExposureDuration: Double {
    return captureDevice.activeMaxExposureDuration.seconds
  }

  var exposureBias: Double {
    return Double(captureDevice.exposureTargetBias)
  }

  var iso: Double {
    return Double(captureDevice.iso)
  }

  func configure(config: CameraControllerConfiguration) -> Promise<Void> {
    return captureDevice.withLock(queue) {
      let device = self.captureDevice

      // `enableLowLightBoost={...}`
      if let enableLowLightBoost = config.enableLowLightBoost {
        guard device.isLowLightBoostSupported else {
          throw RuntimeError.error(
            withMessage: "This CameraDevice does not support `enableLowLightBoost`!")
        }
        device.automaticallyEnablesLowLightBoostWhenAvailable = enableLowLightBoost
      }

      // `enableSmoothAutoFocus={...}`
      if let enableSmoothAutoFocus = config.enableSmoothAutoFocus {
        guard device.isSmoothAutoFocusSupported else {
          throw RuntimeError.error(
            withMessage: "This CameraDevice does not support `enableSmoothAutoFocus`!")
        }
        device.isSmoothAutoFocusEnabled = enableSmoothAutoFocus
      }

      // `enableDistortionCorrection={...}`
      if let enableDistortionCorrection = config.enableDistortionCorrection {
        guard device.isGeometricDistortionCorrectionSupported else {
          throw RuntimeError.error(
            withMessage: "This CameraDevice does not support `enableDistortionCorrection`!")
        }
        device.isGeometricDistortionCorrectionEnabled = enableDistortionCorrection
      }
    }
  }

  func setZoom(zoom: Double) -> Promise<Void> {
    return captureDevice.withLock(queue) {
      // 1. Ensure zoom is within valid range
      let device = self.captureDevice
      guard zoom >= device.minAvailableVideoZoomFactor && zoom <= device.maxAvailableVideoZoomFactor
      else {
        throw RuntimeError.error(
          withMessage:
            "`zoom` is not within the device's `minZoom` and `maxZoom` range! (Received: \(zoom), `device.minZoom`: \(device.minAvailableVideoZoomFactor), `device.maxZoom`: \(device.maxAvailableVideoZoomFactor)"
        )
      }
      // 2. Clamp zoom to the active format's max zoom (might be slightly smaller than device max zoom)
      let clampedForFormat = min(zoom, device.activeFormat.videoMaxZoomFactor)
      // 3. Apply zoom
      device.videoZoomFactor = clampedForFormat
    }
  }

  func focusTo(
    point: any HybridMeteringPointSpec,
    options: FocusOptions
  ) -> Promise<Void> {
    return captureDevice.withLock(queue) { resolve, reject in
      guard let point = point as? HybridMeteringPoint else {
        throw RuntimeError.error(withMessage: "MeteringPoint is not of type `HybridMeteringPoint`!")
      }
      let responsiveness = options.responsiveness ?? .snappy
      let adaptiveness = options.adaptiveness ?? .continuous
      let autoResetAfter = options.autoResetAfter ?? Variant_NullType_Double.second(5.0)
      let modes = options.modes ?? self.getAllSupportedMeteringModes()
      guard !modes.isEmpty else {
        throw RuntimeError.error(withMessage: "MeteringModes cannot be empty!")
      }

      // Cancel any previous tasks
      self.activeMeteringTask?.cancel()
      self.activeMeteringTask = nil

      // Create a new MeteringTask and start the desired metering operations
      let task = MeteringTask(
        device: self.captureDevice,
        adaptiveness: adaptiveness,
        autoReset: autoResetAfter.toAutoReset(),
        on: self.queue)
      // AE
      if modes.contains(.ae) {
        try task.startMeteringAE(
          to: point.toNormalizedPoint(),
          responsiveness: responsiveness)
      }
      // AF
      if modes.contains(.af) {
        try task.startMeteringAF(
          to: point.toNormalizedPoint(),
          responsiveness: responsiveness)
      }
      // AWB
      if modes.contains(.awb) {
        try task.startMeteringAWB(responsiveness: responsiveness)
      }

      // Start listening to updates
      task.startListening(
        onComplete: resolve,
        onError: reject
      )
      self.activeMeteringTask = task

    }
  }

  /**
   * Get a list of all [MeteringMode]s that are supported on this device.
   */
  private func getAllSupportedMeteringModes() -> [MeteringMode] {
    var modes: [MeteringMode] = []
    if captureDevice.isExposurePointOfInterestSupported {
      modes.append(.ae)
    }
    if captureDevice.isFocusPointOfInterestSupported {
      modes.append(.af)
    }
    // White Balance adjusting is always supported, but it's not to a specific point
    modes.append(.awb)
    return modes
  }

  func resetFocus() -> Promise<Void> {
    return captureDevice.withLock(queue) {
      // cancel an ongoing focus operation
      self.activeMeteringTask?.cancel()
      self.activeMeteringTask = nil

      logger.info("Resetting focus...")
      let device = self.captureDevice
      // reset AF to center + continuous tracking
      if device.isFocusPointOfInterestSupported {
        device.focusPointOfInterest = CGPoint(x: 0.5, y: 0.5)
      }
      if device.isFocusModeSupported(.continuousAutoFocus) {
        device.focusMode = .continuousAutoFocus
      }
      // reset AE to center + continuous tracking
      if device.isExposurePointOfInterestSupported {
        device.exposurePointOfInterest = CGPoint(x: 0.5, y: 0.5)
      }
      if device.isExposureModeSupported(.continuousAutoExposure) {
        device.exposureMode = .continuousAutoExposure
      }
      // reset AWB to continuous tracking
      if device.isWhiteBalanceModeSupported(.continuousAutoWhiteBalance) {
        device.whiteBalanceMode = .continuousAutoWhiteBalance
      }
    }
  }

  func addSubjectAreaChangedListener(onSubjectAreaChanged: @escaping () -> Void) -> ListenerSubscription {
    return SubjectAreaMonitor.addObserver(device: captureDevice,
                                          onSubjectAreaDidChange: onSubjectAreaChanged)
  }

  func setTorchMode(mode: TorchMode, strength: Double?) -> Promise<Void> {
    return captureDevice.withLock(queue) {
      // 1. Ensure we have a torch
      if !self.captureDevice.hasTorch {
        if mode == .off {
          // We set it to off, we can ignore that we don't have a torch.
          return
        }
        throw RuntimeError.error(
          withMessage:
            "Tried setting torch mode to `\(mode.stringValue)`, but the device does not have a `torch`!"
        )
      }

      // 2. Set mode to on or off
      switch mode {
      case .off:
        self.captureDevice.torchMode = .off
      case .on:
        if let strength {
          // 3. We have a custom strength level too! Set that
          guard strength >= 0 && strength <= 1 else {
            throw RuntimeError.error(
              withMessage:
                "Torch `strength` is not within 0.0 to 1.0 range! (Received: \(strength))")
          }
          try self.captureDevice.setTorchModeOn(level: Float(strength))
        } else {
          self.captureDevice.torchMode = .on
        }
      }
    }
  }

  func startZoomAnimation(zoom: Double, rate: Double) -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      self.captureDevice.ramp(toVideoZoomFactor: zoom, withRate: Float(rate))

      var observation: NSKeyValueObservation?
      observation = self.captureDevice.observe(
        \.isRampingVideoZoom,
        changeHandler: { _, change in
          if change.oldValue == true && change.newValue == false {
            completion()
            observation?.invalidate()
          }
        })
    }
  }

  func cancelZoomAnimation() -> Promise<Void> {
    return captureDevice.withLock(queue) {
      self.captureDevice.cancelVideoZoomRamp()
    }
  }

  func setExposureBias(exposure: Double) -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      // 1. Ensure exposureBias is within range
      let device = self.captureDevice
      let exposure = Float(exposure)
      guard exposure >= device.minExposureTargetBias && exposure <= device.maxExposureTargetBias
      else {
        throw RuntimeError.error(
          withMessage:
            "`exposure` is not within the device's `minExposureBias` and `maxExposureBias` range! (Received: \(exposure), `device.minExposureBias`: \(device.minExposureTargetBias), `device.maxExposureBias`: \(device.maxExposureTargetBias)"
        )
      }

      // 2. Set Exposure Bias
      device.setExposureTargetBias(Float(exposure)) { _ in
        // 3. Once exposure is set, resolve the Promise
        completion()
      }
    }
  }

  func setFocusLocked(lensPosition: Double) -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      // 1. Ensure values are within valid range
      guard lensPosition >= 0 && lensPosition <= 1 else {
        throw RuntimeError.error(
          withMessage: "`lensPosition` is not within 0.0 to 1.0 range! (Received: \(lensPosition))")
      }

      // 2. Lock Focus
      self.captureDevice.setFocusModeLocked(lensPosition: Float(lensPosition)) { _ in
        // 3. Once it locked, resolve the Promise
        completion()
      }
    }
  }

  func lockCurrentFocus() -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      self.captureDevice.setFocusModeLocked(lensPosition: AVCaptureDevice.currentLensPosition) {
        _ in
        completion()
      }
    }
  }

  var minExposureDuration: Double {
    return captureDevice.activeFormat.minExposureDuration.seconds
  }

  var maxExposureDuration: Double {
    return captureDevice.activeFormat.maxExposureDuration.seconds
  }

  var minISO: Double {
    return Double(captureDevice.activeFormat.minISO)
  }

  var maxISO: Double {
    return Double(captureDevice.activeFormat.maxISO)
  }

  func setExposureLocked(duration: Double, iso: Double) -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      let device = self.captureDevice
      // 1. Ensure values are within valid range
      guard
        duration >= device.activeFormat.minExposureDuration.seconds
          && duration <= device.activeFormat.maxExposureDuration.seconds
      else {
        throw RuntimeError.error(
          withMessage:
            "`duration` is not within the currently selected Format's `minExposureDuration` and `maxExposureDuration` range! (Received: \(duration), `activeFormat.minExposureDuration`: \(device.activeFormat.minExposureDuration.seconds), `activeFormat.maxExposureDuration`: \(device.activeFormat.maxExposureDuration.seconds)"
        )
      }
      let iso = Float(iso)
      guard iso >= device.activeFormat.minISO && iso <= device.activeFormat.maxISO else {
        throw RuntimeError.error(
          withMessage:
            "`iso` is not within the currently selected Format's `minISO` and `maxISO` range! (Received: \(iso), `activeFormat.minISO`: \(device.activeFormat.minISO), `activeFormat.maxISO`: \(device.activeFormat.maxISO)"
        )
      }

      // 2. Set the Exposure Duration/ISO
      let time = CMTime(seconds: duration, preferredTimescale: 1_000_000)
      device.setExposureModeCustom(duration: time, iso: iso) { _ in
        // 3. Once exposure is set, resolve the Promise
        completion()
      }
    }
  }

  func lockCurrentExposure() -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      self.captureDevice.setExposureModeCustom(
        duration: AVCaptureDevice.currentExposureDuration,
        iso: AVCaptureDevice.currentISO
      ) { _ in
        completion()
      }
    }
  }

  func convertWhiteBalanceTemperatureAndTintValues(
    whiteBalanceTemperatureAndTint: WhiteBalanceTemperatureAndTint
  ) throws -> WhiteBalanceGains {
    // 1. Convert our type to the AVFoundation type
    let tintValues = AVCaptureDevice.WhiteBalanceTemperatureAndTintValues(
      temperature: Float(whiteBalanceTemperatureAndTint.temperature),
      tint: Float(whiteBalanceTemperatureAndTint.tint))
    // 2. Do the actual device conversion
    let gains = self.captureDevice.deviceWhiteBalanceGains(for: tintValues)
    // 3. Convert the returned AVFoundation gains to our type
    return WhiteBalanceGains(
      redGain: Double(gains.redGain),
      blueGain: Double(gains.blueGain),
      greenGain: Double(gains.greenGain))
  }

  func setWhiteBalanceLocked(whiteBalanceGains: WhiteBalanceGains) -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      // 1. Ensure values are within valid range
      let device = self.captureDevice
      let redGain = Float(whiteBalanceGains.redGain)
      let greenGain = Float(whiteBalanceGains.greenGain)
      let blueGain = Float(whiteBalanceGains.blueGain)
      guard redGain >= 1 && redGain <= device.maxWhiteBalanceGain else {
        throw RuntimeError.error(
          withMessage:
            "`whiteBalanceGains.redGain` is not within the range of `1` to `device.maxWhiteBalanceGain`! (Received: \(redGain), min: 1.0, `device.maxWhiteBalanceGain`: \(device.maxWhiteBalanceGain)"
        )
      }
      guard greenGain >= 1 && greenGain <= device.maxWhiteBalanceGain else {
        throw RuntimeError.error(
          withMessage:
            "`whiteBalanceGains.greenGain` is not within the range of `1` to `device.maxWhiteBalanceGain`! (Received: \(greenGain), min: 1.0, `device.maxWhiteBalanceGain`: \(device.maxWhiteBalanceGain)"
        )
      }
      guard blueGain >= 1 && blueGain <= device.maxWhiteBalanceGain else {
        throw RuntimeError.error(
          withMessage:
            "`whiteBalanceGains.blueGain` is not within the range of `1` to `device.maxWhiteBalanceGain`! (Received: \(blueGain), min: 1.0, `device.maxWhiteBalanceGain`: \(device.maxWhiteBalanceGain)"
        )
      }

      // 2. Lock White Balance Gains
      let values = AVCaptureDevice.WhiteBalanceGains(
        redGain: redGain,
        greenGain: greenGain,
        blueGain: blueGain)
      device.setWhiteBalanceModeLocked(with: values) { _ in
        // 3. Once White Balance is locked, resolve the Promise
        completion()
      }
    }
  }

  func lockCurrentWhiteBalance() -> Promise<Void> {
    return captureDevice.withLock(queue) { completion in
      self.captureDevice.setWhiteBalanceModeLocked(with: AVCaptureDevice.currentWhiteBalanceGains) {
        _ in
        completion()
      }
    }
  }
}

File: packages/react-native-vision-camera/android/src/main/java/com/margelo/nitro/camera/hybrids/HybridCameraController.kt
package com.margelo.nitro.camera.hybrids

import android.animation.Animator
import android.animation.ValueAnimator
import androidx.camera.core.Camera
import androidx.camera.core.CameraState
import androidx.camera.core.FocusMeteringAction
import androidx.camera.core.LowLightBoostState
import androidx.camera.core.MeteringPoint
import androidx.camera.core.TorchState
import com.margelo.nitro.camera.CameraControllerConfiguration
import com.margelo.nitro.camera.ExposureMode
import com.margelo.nitro.camera.FocusMode
import com.margelo.nitro.camera.FocusOptions
import com.margelo.nitro.camera.HybridCameraControllerSpec
import com.margelo.nitro.camera.HybridCameraDeviceSpec
import com.margelo.nitro.camera.HybridMeteringPointSpec
import com.margelo.nitro.camera.MeteringMode
import com.margelo.nitro.camera.TorchMode
import com.margelo.nitro.camera.Variant_NullType_Double
import com.margelo.nitro.camera.WhiteBalanceGains
import com.margelo.nitro.camera.WhiteBalanceMode
import com.margelo.nitro.camera.WhiteBalanceTemperatureAndTint
import com.margelo.nitro.camera.extensions.await
import com.margelo.nitro.camera.extensions.converters.fromTorchState
import com.margelo.nitro.camera.extensions.converters.toMeteringMode
import com.margelo.nitro.camera.hybrids.inputs.HybridCameraDevice
import com.margelo.nitro.camera.hybrids.metering.HybridMeteringPoint
import com.margelo.nitro.core.Promise
import com.margelo.nitro.core.resolve
import com.margelo.nitro.core.resolved
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

class HybridCameraController(
  val camera: Camera,
) : HybridCameraControllerSpec() {
  override val device: HybridCameraDeviceSpec = HybridCameraDevice(camera.cameraInfo)
  private val cameraState: CameraState?
    get() = camera.cameraInfo.cameraState.value
  private var zoomAnimator: ValueAnimator? = null

  override val isConnected: Boolean
    get() = cameraState?.type == CameraState.Type.OPEN

  override val isSuspended: Boolean
    get() = cameraState?.error?.code == CameraState.ERROR_DO_NOT_DISTURB_MODE_ENABLED

  override val isUsedByAnotherApp: Boolean
    get() = cameraState?.error?.code == CameraState.ERROR_CAMERA_IN_USE

  override val zoom: Double
    get() =
      camera.cameraInfo.zoomState.value
        ?.zoomRatio
        ?.toDouble() ?: 1.0
  override val minZoom: Double
    get() =
      camera.cameraInfo.zoomState.value
        ?.minZoomRatio
        ?.toDouble() ?: 1.0
  override val maxZoom: Double
    get() =
      camera.cameraInfo.zoomState.value
        ?.maxZoomRatio
        ?.toDouble() ?: 1.0
  override val displayableZoomFactor: Double
    get() = zoom * camera.cameraInfo.intrinsicZoomRatio

  override val torchStrength: Double
    get() =
      camera.cameraInfo.torchStrengthLevel.value
        ?.toDouble() ?: 1.0
  override val torchMode: TorchMode
    get() = TorchMode.fromTorchState(camera.cameraInfo.torchState.value ?: TorchState.OFF)

  override val exposureBias: Double
    get() =
      camera.cameraInfo.exposureState.exposureCompensationIndex
        .toDouble()

  override val isLowLightBoostEnabled: Boolean
    get() = camera.cameraInfo.lowLightBoostState.value != LowLightBoostState.OFF
  override val isDistortionCorrectionEnabled: Boolean
    get() = false
  override val isSmoothAutoFocusEnabled: Boolean
    get() = false

  // TODO: CameraX only supports continuous AE/AF/AWB
  override val focusMode: FocusMode
    get() = FocusMode.CONTINUOUS_AUTO_FOCUS
  override val exposureMode: ExposureMode
    get() = ExposureMode.CONTINUOUS_AUTO_EXPOSURE
  override val whiteBalanceMode: WhiteBalanceMode
    get() = WhiteBalanceMode.CONTINUOUS_AUTO_WHITE_BALANCE

  override fun configure(config: CameraControllerConfiguration): Promise<Unit> {
    return Promise.async {
      val futures = mutableListOf<Future<*>>()

      // `enableLowLightBoost={...}`
      if (config.enableLowLightBoost != null) {
        futures += camera.cameraControl.enableLowLightBoostAsync(config.enableLowLightBoost)
      }

      // `enableSmoothAutoFocus={...}`
      if (config.enableSmoothAutoFocus != null) {
        // TODO: Implement enableSmoothAutoFocus changing when CameraX supports it
      }
      // `enableDistortionCorrection={...}`
      if (config.enableDistortionCorrection != null) {
        // TODO: Implement enableDistortionCorrection changing when CameraX supports it
      }

      futures.forEach { it.await() }
    }
  }

  override fun setZoom(zoom: Double): Promise<Unit> {
    return Promise.async {
      val zoomState = camera.cameraInfo.zoomState.value
      val zoom = zoom.toFloat()
      if (zoomState != null) {
        if (zoom < zoomState.minZoomRatio || zoom > zoomState.maxZoomRatio) {
          throw Error(
            "`zoom` is out of range! Expected value within " +
              "${zoomState.minZoomRatio}...${zoomState.maxZoomRatio}, received $zoom.",
          )
        }
      }
      camera.cameraControl
        .setZoomRatio(zoom)
        .await()
    }
  }

  override fun focusTo(
    point: HybridMeteringPointSpec,
    options: FocusOptions,
  ): Promise<Unit> {
    return Promise.async {
      val point =
        point as? HybridMeteringPoint
          ?: throw Error("Point is not of type `HybridMeteringPoint`!")
      val meteringPoint = point.meteringPoint
      val meteringMode = (options.modes ?: getAllSupportedMeteringModes(meteringPoint)).toMeteringMode()
      val autoResetAfter = options.autoResetAfter ?: Variant_NullType_Double.create(5.0)

      val focusAction =
        FocusMeteringAction.Builder(meteringPoint, meteringMode).also { action ->
          autoResetAfter.match(
            { _ -> action.disableAutoCancel() },
            { duration -> action.setAutoCancelDuration(duration.toLong(), TimeUnit.SECONDS) },
          )
        }
      camera.cameraControl
        .startFocusAndMetering(focusAction.build())
        .await()
    }
  }

  /**
   * Get a list of all [MeteringMode]s that are supported on this device.
   */
  private fun getAllSupportedMeteringModes(point: MeteringPoint): Array<MeteringMode> {
    val allModes = arrayOf(MeteringMode.AE, MeteringMode.AF, MeteringMode.AWB)
    return allModes
      .filter { mode ->
        val action = FocusMeteringAction.Builder(point, mode.toMeteringMode()).build()
        camera.cameraInfo.isFocusMeteringSupported(action)
      }.toTypedArray()
  }

  override fun resetFocus(): Promise<Unit> {
    return Promise.async {
      camera.cameraControl
        .cancelFocusAndMetering()
        .await()
    }
  }

  override fun addSubjectAreaChangedListener(onSubjectAreaChanged: (() -> Unit)?) {
    // TODO: CameraX does not yet support subject area changed listeners.
    //       Feature Request: https://issuetracker.google.com/issues/505643406
  }

  override fun setExposureBias(exposure: Double): Promise<Unit> {
    return Promise.async {
      camera.cameraControl
        .setExposureCompensationIndex(exposure.toInt())
        .await()
    }
  }

  override fun setTorchMode(
    mode: TorchMode,
    strength: Double?,
  ): Promise<Unit> {
    return Promise.async {
      when (mode) {
        TorchMode.OFF -> {
          camera.cameraControl
            .enableTorch(false)
            .await()
        }
        TorchMode.ON -> {
          if (strength != null) {
            require(strength in 0.0..1.0) {
              "Torch `strength` is not within 0.0 to 1.0 range! (Received: $strength)"
            }
            val normalizedStrength = 1 + (strength * camera.cameraInfo.maxTorchStrengthLevel)
            camera.cameraControl
              .setTorchStrengthLevel(normalizedStrength.toInt())
              .await()
            camera.cameraControl
              .enableTorch(true)
              .await()
          } else {
            camera.cameraControl
              .enableTorch(true)
              .await()
          }
        }
      }
    }
  }

  override fun startZoomAnimation(
    zoom: Double,
    rate: Double,
  ): Promise<Unit> {
    val promise = Promise<Unit>()

    zoomAnimator?.cancel()

    val from = this.zoom.toFloat()
    val to = zoom.toFloat()
    zoomAnimator =
      ValueAnimator.ofFloat(from, to).apply {
        duration = rate.toLong()
        addUpdateListener {
          val currentZoomRatio = (animatedValue as Float)
          camera.cameraControl.setZoomRatio(currentZoomRatio)
        }
        addListener(
          object : Animator.AnimatorListener {
            override fun onAnimationCancel(p0: Animator) {
              val error = Error("The Zoom animation was canceled!")
              promise.reject(error)
            }

            override fun onAnimationEnd(animator: Animator) {
              val isCanceled = !isRunning && isStarted && animatedFraction != 1f
              if (!isCanceled) {
                promise.resolve()
              }
            }

            override fun onAnimationRepeat(animator: Animator) = Unit

            override fun onAnimationStart(animator: Animator) = Unit
          },
        )
        start()
      }

    return promise
  }

  override fun cancelZoomAnimation(): Promise<Unit> {
    zoomAnimator?.cancel()
    zoomAnimator = null
    return Promise.resolved()
  }

  // TODO: Once CameraX implements locking Focus to a custom lens position, implement this:
  override val lensPosition: Double
    get() = 0.0

  override fun setFocusLocked(lensPosition: Double): Promise<Unit> {
    throw Error(
      "Setting Focus Lens Position is not yet supported on Android! " +
        "You can use AF focus (`focusTo(..., ['AF'])`) instead.",
    )
  }

  override fun lockCurrentFocus(): Promise<Unit> {
    throw Error(
      "Setting Focus Lens Position is not yet supported on Android! " +
        "You can use AF focus (`focusTo(..., ['AF'])`) instead.",
    )
  }

  // TODO: Once CameraX implements locking Exposure to custom duration/ISO, implement this:
  override val minExposureDuration: Double
    get() = 0.0
  override val maxExposureDuration: Double
    get() = 0.0
  override val minISO: Double
    get() = 0.0
  override val maxISO: Double
    get() = 0.0
  override val iso: Double
    get() = 0.0
  override val exposureDuration: Double
    get() = 0.0

  override fun setExposureLocked(
    duration: Double,
    iso: Double,
  ): Promise<Unit> {
    throw Error(
      "Locking Exposure to manual duration/ISO values is not yet supported on Android! " +
        "You can use Exposure Bias (`setExposureBias(...)`) or AE focus (`focusTo(..., ['AE'])`) instead.",
    )
  }

  override fun lockCurrentExposure(): Promise<Unit> {
    throw Error(
      "Locking Exposure to manual duration/ISO values is not yet supported on Android! " +
        "You can use Exposure Bias (`setExposureBias(...)`) or AE focus (`focusTo(..., ['AE'])`) instead.",
    )
  }

  // TODO: Once CameraX implements locking AWB to custom White Balance Gains, implement this:
  override val whiteBalanceGains: WhiteBalanceGains
    get() = WhiteBalanceGains(0.0, 0.0, 0.0)

  override fun convertWhiteBalanceTemperatureAndTintValues(whiteBalanceTemperatureAndTint: WhiteBalanceTemperatureAndTint): WhiteBalanceGains {
    throw Error(
      "Locking White Balance Gains is not yet supported on Android! " +
        "You can use AWB focus (`focusTo(..., ['AWB'])`) instead.",
    )
  }

  override fun setWhiteBalanceLocked(whiteBalanceGains: WhiteBalanceGains): Promise<Unit> {
    throw Error(
      "Locking White Balance Gains is not yet supported on Android! " +
        "You can use AWB focus (`focusTo(..., ['AWB'])`) instead.",
    )
  }

  override fun lockCurrentWhiteBalance(): Promise<Unit> {
    throw Error(
      "Locking White Balance Gains is not yet supported on Android! " +
        "You can use AWB focus (`focusTo(..., ['AWB'])`) instead.",
    )
  }
}
